nixos/modules/services/networking/sozu.nix

```diff
@@ -0,0 +1,479 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.services.sozu;
+  format = pkgs.formats.toml {};
+
+  globalSettings = format.generate "global.toml" cfg.global;
+  optionalSettings = format.generate "optional.toml" cfg.optional;
+in
+{
+  options.services.sozu = {
+    enable = mkEnableOption "Sozu server";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.sozu;
+      description = "Sozu package to use.";
+    };
+
+    configFile = mkOption {
+      type = types.either types.str types.path;
+      default = pkgs.writeText "config.toml" ''
+        ${fileContents globalSettings}
+
+        ${fileContents optionalSettings}
+      '';
+      description = ''
+        Override the configuration file used by Sozu.
+        By default, NixOS generates on automatically.
+      '';
+    };
+
+    global = mkOption {
+      default = {};
+      description = ''
+        Global settings that will be parsed into Sozu's <literal>config.toml</literal>.
+        Documentation <link xlink:href="https://github.com/sozu-proxy/sozu/blob/master/doc/configure.md">here</link>.
+      '';
+      type = types.submodule {
+
+        options = {
```

Comment:
Are all of these options required? It seems like some people might **not** want to include some of these options into their configuration from the description. I would suggest you go for a more minimal approach here (less typed options) and fill in values in the `config` if needed.

Reply:
Let me know if it's good now.

Reply:
I was thinking something a little bit more simple like [this](https://github.com/aanderse/nixpkgs/blob/cfad31a7310d1b3d1723f53598fd65f3a069bfd9/nixos/modules/services/networking/sozu.nix).

Note that you could still using a `freeformType` to provide restrictions around some options if needed, though it is likely just overhead that doesn't provide as much value in this case... most of the general configuration options are pretty self explanatory.

Reply:
Also worth mentioning is the removal of the `configFile` option. This complicates things because `restartTriggers` stops working and the module becomes pretty much worthless. You can't reference anything out of `settings` anymore because it might be wrong.

I think the solution to having users use `toml` natively is `builtins.fromToml`, perhaps. Maybe `services.sozu.settings = builtins.fromToml /path/to/my/config.toml;`.